cscs-checks/microbenchmarks/gpu/pointer_chase/src/linked_list.hpp

```diff
@@ -0,0 +1,383 @@
+  if (ptr == ptrs[0])
+  {
+    printf("This is some data dependency that will never be executed.");
+  }
+#endif
+
+  // Join the tail with the head (just for the data dependency).
+  if (ptr->next == nullptr)
+  {
+    ptr->next = &(buffer[headIndex]);
+  }
+
+}
+
+
+/*
+ * List structure definitions
+ */
+
+struct List
+{
+  /*
+   * Contains the buffer where the list is stored, the index in this buffer where the head
+   * of the list is, the buffer size, and the stride in between nodes (this last one is only
+   * meaningful if the list is not initialised as random).
+   *
+   * The member functions are:
+   *  - info: prints the list details.
+   *  - initialize: populate the buffer with the list nodes.
+   *  - traverse: simple list traversal.
+   *  - timed_traverse: traverses the list and measures the number of cycles per node jump.
+   */
+
+  Node * buffer = nullptr;
+  uint32_t headIndex = 0;
+  uint32_t * timer = nullptr;
+  uint32_t * d_timer = nullptr;
+  size_t buffSize;
+  size_t stride;
+
+  List(size_t bSize, size_t st) : buffSize(bSize), stride(st)
+  {
+    // Allocate the buffers to store the timings measured in the kernel
+    timer = new uint32_t[NODES];
+    XMalloc((void**)&d_timer, sizeof(uint32_t)*(NODES));
+  };
+
+  virtual ~List()
+  {
+    XFree(d_timer);
+  }
+
+  void info(size_t n, size_t buffSize)
+  {
+    printf("Creating Linked list:\n");
+    printf(" - Node size: %lu\n", sizeof(Node));
+    printf(" - Number of nodes: %lu:\n", n);
+    printf(" - Total buffer size: %10.2f MB:\n", float(sizeof(Node)*buffSize)/1024.0/1024);
+    XDeviceSynchronize();
+  }
+
+  void initialize(int mode=0)
+  {
+    /*
+     * mode 0 initializes the list as serial.
+     * mode 1 initializes the list in a random order.
+     */
+
+    if (mode < 0 || mode > 1)
+    {
+      printf("Unknown list initialization scheme. Defaulting back to 0.");
+      mode = 0;
+    }
+
+    if (mode == 0)
+    {
+      initialize_list<<<1,1>>>(buffer, stride);
+      XDeviceSynchronize();
+    }
+    else
+    {
+      // Random number engine.
+      std::mt19937_64 rng;
+      uint64_t timeSeed = std::chrono::high_resolution_clock::now().time_since_epoch().count();
+      std::seed_seq ss{uint32_t(timeSeed & 0xffffffff), uint32_t(timeSeed>>32)};
+      rng.seed(ss);
+      std::uniform_real_distribution<double> unif(0, 1);
+
+      uint32_t * nodeIndices = (uint32_t*)malloc(sizeof(uint32_t)*NODES);
+      // Create set to keep track of the assigned indices.
+      std::set<uint32_t> s = {};
+      for (int i = 0; i < NODES; i++)
+      {
+        // Get a random index.
+        uint32_t currentIndex = (uint32_t)(unif(rng)*buffSize);
+
+        // If already present in the set, find another alternative index.
+        while (s.find(currentIndex) != s.end())
+        {
+          if (currentIndex < NODES-1)
+          {
+            currentIndex++;
+          }
+          else
+          {
+            currentIndex = 0;
+          }
+        }
```

Comment:
To generate a random permutation, you usually start from the sorted sequence 0...NODES, then apply a random transposition to each element. `std::iota` plus `std::shuffle` does quite nicely:
```suggestion
            std::random_device rd;
            std::mt19937_64 gen(rd());
            uint32_t * nodeIndices = (uint32_t*)malloc(sizeof(uint32_t)*NODES);
            std::shuffle(nodeIndices, nodeIndices + NODES, gen);
```